### PR TITLE
Introduce a notion of a simple collator

### PR DESCRIPTION
**Pull request by pepyakin**
The reason for this is to support the simplest collator ever: adder-collator.

I recommend looking into each commit independently.

### CODE REVIEW

File: src/spec.js [hunk 1]
```diff
@@ -9,7 +9,12 @@ function nameCase(string) {
 // Remove all existing keys from `session.keys`
 export function clearAuthorities(spec) {
 	let rawdata = fs.readFileSync(spec);
-	let chainSpec = JSON.parse(rawdata);
+	let chainSpec;
+	try {
+		chainSpec = JSON.parse(rawdata);
+	} catch {
+		console.error("failed to parse the chain spec");
```
Comment (bkchr) [inline src/spec.js:16]: Do we need to return here?

File: src/spawn.js [hunk 2]
```diff
@@ -18,10 +19,9 @@ export async function generateChainSpec(bin, chain) {
...
 
-		p['spec'].stdout.on('data', function (chunk) {
-			let message = chunk.toString();
-			spec.write(message);
-		});
+		// `pipe` since it deals with flushing and  we need to guarantee that the data is flushed
+		// before we resolve the promise.
+		p['spec'].stdout.pipe(spec);
 
 		p['spec'].stderr.on('data', function (chunk) {
```
Comment (bkchr) [inline src/spawn.js:26]: Can we not pipe this as well to stderr?

File: src/index.js [hunk 3]
```diff
@@ -63,37 +65,61 @@ async function main() {
...
-		// Connect to the first relay chain node to submit the extrinsic.
-		let relayChainApi = await connect(config.relaychain.nodes[0].wsPort, config.types);
-		await registerParachain(relayChainApi, id, wasm, header);
+		let genesisState
+		let genesisWasm
+		try {
+			genesisState = await exportGenesisState(bin, null);
+			genesisWasm = await exportGenesisWasm(bin, null);
+		} catch (err) {
+			console.error(err);
```
Comment (bkchr) [inline src/index.js:118]: exit?

File: src/spawn.js (hunk 2 above)
Comment (pepyakin) [reply src/spawn.js:26]: or, perhaps, just inherit stderr?

File: src/index.js (hunk 3 above)
Comment (pepyakin) [reply src/index.js:118]: ah weird, my push didn't go through

File: src/spawn.js (hunk 2 above)
Comment (bkchr) [reply src/spawn.js:26]: If that works, even better :)

**Review by bkchr — APPROVED** [review]
Looks good as far as I can see. However, my js is really bad..

I think some update of the README and an example config would be nice before merging this :)

**Comment by pepyakin** [comment]
Good ideas right there

However, I've noticed that somehow this got cumulus broken somehow :< Investigating

File: src/spawn.js [hunk 4]
```diff
@@ -92,38 +84,52 @@ export function startNode(bin, name, wsPort, port, spec, flags) {
...
+	}
+	return stdout.trim()
+}
+
+/// Export the genesis state aka genesis head.
+export async function exportGenesisState(bin, id, chain) {
+	let args = [
+		"export-genesis-state",
+	]
+	if (id) {
```
Comment (bkchr) [inline src/spawn.js:114]: Javascript :heart: 

Ohh boy, I'm loving statically typed  languages sooooo much xD

**Pull request merged**